File: backend/domain/services/statistics_service.py
from datetime import date, timedelta
from typing import List

from .account_service import AccountService
from .transaction_service import TransactionService
from ..models import KeyValue
from ..models import PeriodType
from ..models import Summary
from ..search_request import FilterRequest, FilterOperator
from ...mapping import Mapper
from ...modules.depynject import injectable


@injectable()
class StatisticsService:
    """
    The statistics service class that defines all business operations.
    """

    def __init__(self,
                 object_mapper: Mapper,
                 transaction_service: TransactionService,
                 account_service: AccountService
                 ) -> None:
        """Constructor

        :param object_mapper: the object mapper
        :param transaction_service: the transaction service
        :param account_service: the account service
        """
        self.__mapper = object_mapper
        self.__transaction_service = transaction_service
        self.__account_service = account_service

    def get_total_over_period(self,
                              period: PeriodType,
                              filter_request: FilterRequest
                              ) -> List[KeyValue]:
        """Gets the total of all transaction for a given period.

        :param period: the period
        :param filter_request: the filter request
        :return: the list of (key, value) results
        """
        return self.__transaction_service.get_total_over_period(period, filter_request)

    def get_evolution_over_period(self,
                                  period: PeriodType,
                                  account_ids: List[int],
                                  date_from: date,
                                  date_to: date,
                                  filter_request: FilterRequest
                                  ) -> List[KeyValue]:
        """Gets the evolution over a given year.

        :param period: the period
        :param account_ids: the account ids
        :param date_from: the from date
        :param date_to: the to date
        :param filter_request: the filter request
        :return: the list of (key, value) results
        """
        filters = FilterRequest.all(
            FilterRequest.of('date_value', date_from, FilterOperator.GE),
            FilterRequest.of('date_value', date_to, FilterOperator.LT),
            FilterRequest.of('account_id', account_ids, FilterOperator.IN),
            filter_request
        )

        # Retrieving all entries
        entries = self.__transaction_service.get_total_over_period(period, filters)

        # Calculating the start amount
        start_amount = 0
        for acc_id in account_ids:
            account_total = self.__account_service.get_account_total(acc_id, date_from)
            if account_total is not None:
                start_amount = start_amount + account_total

        # Calculating the start date
        key: str = ''
        if period == PeriodType.DAY:
            first_date = date_from - timedelta(days=1)
            key = first_date.strftime("%Y-%m-%d")
        if period == PeriodType.MONTH:
            first_date = date_from - timedelta(days=30)
            key = first_date.strftime("%Y-%m")
        if period == PeriodType.QUARTER:
            first_date = date_from - timedelta(days=90)
            key = first_date.strftime("%Y") + '-' + str((first_date.month + 2) / 3)
        if period == PeriodType.YEAR:
            first_date = date_from - timedelta(days=365)
            key = first_date.strftime("%Y")
        values = [KeyValue(key, start_amount)]

        # Re-calculating the evolution with the start amount
        for e in entries:
            if e.value is not None:
                start_amount = start_amount + e.value
            values.append(KeyValue(e.key, start_amount))
        return values

    def get_summary(self,
                    account_ids: List[int],
                    date_from: date,
                    date_to: date,
                    filter_request: FilterRequest
                    ) -> Summary:
        """Gets the summary for the provided filters.

        :param account_ids: the account ids
        :param date_from: the from date
        :param date_to: the to date
        :param filter_request: the filter request
        :return: the summary
        """
        filters = FilterRequest.all(
            FilterRequest.of('date_value', date_from, FilterOperator.GE),
            FilterRequest.of('date_value', date_to, FilterOperator.LT),
            FilterRequest.of('account_id', account_ids, FilterOperator.IN),
            filter_request
        )
        credit_filters = FilterRequest.all(
            FilterRequest.of('amount', 0, FilterOperator.GE),
            filters
        )
        debit_filters = FilterRequest.all(
            FilterRequest.of('amount', 0, FilterOperator.LT),
            filters
        )

        # Retrieving all entries
        total_debit = self.__transaction_service.get_total(debit_filters)
        total_credit = self.__transaction_service.get_total(credit_filters)

        # Calculating the start amounts
        amount_start = 0
        amount_end = 0
        for acc_id in account_ids:
            amount_start = amount_start + self.__account_service.get_account_total(acc_id, date_from)
            amount_end = amount_end + self.__account_service.get_account_total(acc_id, date_to)

        return Summary(
            amount_start,
            amount_end,
            total_credit,
            total_debit
        )

File: frontend/src/app/core/services/domain/statistics.service.ts
import { Injectable } from '@angular/core';
import { RqlService } from '../rql.service';
import { StatisticsRestService } from '../rest/statistics-rest.service';
import { FilterOperator, FilterRequest } from '../../models/rql.models';
import { Observable } from 'rxjs';
import { CompositeKeyValue, KeyValue, Summary } from '../../models/api.models';

/**
 * The statistics service
 */
@Injectable()
export class StatisticsService {

  /**
   * Instantiates the service.
   *
   * @param statisticsRestService the statistics rest service
   * @param rqlService the RQL service
   */
  constructor(private statisticsRestService: StatisticsRestService,
              private rqlService: RqlService
  ) {}

  /**
   * Gets the summary of a given list of accounts for a given period (year on month).
   *
   * @param accounts the list of accounts
   * @param year the year
   * @param month the optional month
   */
  getSummary(accounts: number[], year: number, month?: number): Observable<Summary> {
    // Building start and end date
    let dateFrom: Date;
    let dateTo: Date;
    if (month == null) {
      dateFrom = new Date(year, 0, 1);
      dateTo = new Date(year + 1, 0, 1);
    } else {
      dateFrom = new Date(year, month - 1, 1);
      if (month == 12) {
        dateTo = new Date(year + 1, 0, 1);
      } else {
        dateTo = new Date(year, month, 1);
      }
    }

    return this.statisticsRestService.getSummary(dateFrom, dateTo, accounts, undefined);
  }

  /**
   * Gets the aggregation by labels of a given list of accounts.
   *
   * @param year the year
   * @param accounts the list of accounts
   * @param labels the labels to filter
   * @param credit filter credit transactions if true, filters debit transactions if false
   */
  getAggregation(year: number, accounts: number[], labels: number[], credit: boolean): Observable<KeyValue[]> {
    // Building start and end date
    const dateFrom = this.rqlService.formatDate(new Date(year, 0, 1));
    const dateTo = this.rqlService.formatDate(new Date(year + 1, 0, 1));

    // Adding date and amount filters
    let filter = FilterRequest.all(
      FilterRequest.of('dateValue', dateFrom, FilterOperator.GE),
      FilterRequest.of('dateValue', dateTo, FilterOperator.LT),
      FilterRequest.of('amount', 0, credit ? FilterOperator.GE : FilterOperator.LT)
    );

    // Adding the account filters
    if (accounts && accounts.length > 0) {
      const accountList = this.rqlService.formatList(accounts);
      filter = FilterRequest.all(
        FilterRequest.of('accountId', accountList, FilterOperator.IN),
        filter
      );
    }

    // Adding the label filters
    if (labels && labels.length > 0) {
      const accountList = this.rqlService.formatList(labels);
      filter = FilterRequest.all(
        FilterRequest.of('labelId', accountList, FilterOperator.IN),
        filter
      );
    }

    return this.statisticsRestService.getAggregation('MONTH', filter)
  }

  /**
   * Gets the evolution for a given list of accounts.
   *
   * @param year the year of the evolution
   * @param accounts the list of accounts
   */
  getEvolution(year: number, accounts: number[],): Observable<KeyValue[]> {
    const dateFrom = new Date(year, 0, 1);
    const dateTo = new Date(year + 1, 0, 1);

    return this.statisticsRestService.getEvolution('MONTH', dateFrom, dateTo, accounts, undefined);
  }

  /**
   * Gets the repartition for a given list of accounts.
   *
   * @param year the year
   * @param month the month
   * @param accounts the list of accounts
   */
  getRepartition(year: number, month: number, accounts: number[]): Observable<KeyValue[]> {
    // Building start and end date
    const dateFrom = this.rqlService.formatDate(new Date(year, month - 1, 1));
    let dateTo: string;
    if (month == 12) {
      dateTo = this.rqlService.formatDate(new Date(year + 1, 0, 1));
    } else {
      dateTo = this.rqlService.formatDate(new Date(year, month, 1));
    }

    // Adding date and amount filters
    let filter = FilterRequest.all(
      FilterRequest.of('dateValue', dateFrom, FilterOperator.GE),
      FilterRequest.of('dateValue', dateTo, FilterOperator.LT)
    );

    // Adding the account filters
    if (accounts && accounts.length > 0) {
      const accountList = this.rqlService.formatList(accounts);
      filter = FilterRequest.all(
        FilterRequest.of('accountId', accountList, FilterOperator.IN),
        filter
      );
    }

    return this.statisticsRestService.getRepartition(filter);
  }

  /**
   * Gets the analytics for a given list of accounts.
   *
   * @param year the year
   * @param period the aggregation period
   * @param categoryType the category type to filter on
   * @param accounts the list of accounts
   */
  getAnalytics(year: number, period: string, categoryType: string, accounts: number[]): Observable<CompositeKeyValue[]> {
    // Building start and end date
    const dateFrom = this.rqlService.formatDate(new Date(year, 0, 1));
    const dateTo = this.rqlService.formatDate(new Date(year + 1, 0, 1));

    // Adding date and amount filters
    let filter = FilterRequest.all(
      FilterRequest.of('dateValue', dateFrom, FilterOperator.GE),
      FilterRequest.of('dateValue', dateTo, FilterOperator.LT),
      FilterRequest.of('categoryType', categoryType, FilterOperator.EQ)
    );

    // Adding the account filters
    if (accounts && accounts.length > 0) {
      const accountList = this.rqlService.formatList(accounts);
      filter = FilterRequest.all(
        FilterRequest.of('accountId', accountList, FilterOperator.IN),
        filter
      );
    }

    return this.statisticsRestService.getAnalytics(period, filter);
  }

  /**
   * Gets the analytics details for a given list of accounts.
   *
   * @param year the year
   * @param categoryType the category type to filter
   * @param accounts the list of accounts
   */
  getAnalyticsDetails(year: number, categoryType: string, accounts: number[]): Observable<CompositeKeyValue[]> {
    // Building start and end date
    const dateFrom = this.rqlService.formatDate(new Date(year, 0, 1));
    const dateTo = this.rqlService.formatDate(new Date(year + 1, 0, 1));

    // Adding date and amount filters
    let filter = FilterRequest.all(
      FilterRequest.of('dateValue', dateFrom, FilterOperator.GE),
      FilterRequest.of('dateValue', dateTo, FilterOperator.LT),
      FilterRequest.of('categoryType', categoryType, FilterOperator.EQ)
    );

    // Adding the account filters
    if (accounts && accounts.length > 0) {
      const accountList = this.rqlService.formatList(accounts);
      filter = FilterRequest.all(
        FilterRequest.of('accountId', accountList, FilterOperator.IN),
        filter
      );
    }

    return this.statisticsRestService.getAnalyticsDetails(filter);
  }

}
